### PR TITLE
[Routing] Allow using UTF-8 parameter names

### LINKED ISSUE #41909: Route not found with UTF-8 parameter name
**Symfony version(s) affected**: 5.3.2 (at least)

**Description**
Routes with an UTF-8 parameter name can't be found. 

**How to reproduce**
In YAML
```yaml
informes:
    path: /informes/{año}
    controller: App\Controller\InformesController::anual
```
or attributes
```php
#[Route("/informes/{año}", name: "informes")]
public function anual(string $año)
{
    dd($año);
}
```
**Possible Solution**
I don't know, I suppose that is a regression because I remember using this on Symfony 2 many years ago.

**Additional context**
I have tried with YAML, annotations and attributes and only in the last 5.3 version. 
Any route with UTF-8 characters works fine, only fails with UTF-8 parameters name.
The `routing.yaml` have 
```yaml
framework:
    router:
        utf8: true
```
And add `utf8: true` to route options doesn't work either

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 6.1
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #41909
| License       | MIT
| Doc PR        | -

### CODE REVIEW

File: src/Symfony/Component/Routing/Tests/Matcher/UrlMatcherTest.php [hunk 1]
```diff
@@ -941,6 +941,21 @@ public function testRestrictiveTrailingRequirementWithStaticRouteAfter()
...
+    public function testUtf8VarName()
+    {
+        $collection = new RouteCollection();
+        $collection->add('foo', new Route('/foo/{bär}', [], [], ['utf8' => true]));
+        $matcher = $this->getUrlMatcher($collection);
+
+        try {
+            $matcher->match('/no-match');
+            $this->fail();
+        } catch (ResourceNotFoundException $e) {
```
Comment (fancyweb) [inline src/Symfony/Component/Routing/Tests/Matcher/UrlMatcherTest.php]:
```suggestion
        } catch (ResourceNotFoundException) {
```

**Review by fancyweb — APPROVED** [review]
Shouldn't we add a changelog entry?

**Comment by nicolas-grekas** [comment]
Updated, thanks for the review.

**Comment by GromNaN** [comment]
A test could be added to `RouteTest` and `RouteCompilerTest` to cover this feature.

**Comment by nicolas-grekas** [comment]
This is already covered by the current test to me.

**Comment by stof** [comment]
The change in `extractInlineDefaultsAndRequirements` is not covered by tests, as you don't use an UTF-8 name on a parameter using inline config (try reverting that part of the diff and see that tests won't fail).
and changes in places computing `$hasTrailingVar` might need a test using a route using a trailing optional parameter using an UTF-8 name to be covered in a way preventing regression.

**Comment by nicolas-grekas** [comment]
Tests added @stof

**Review by GromNaN — APPROVED** [review]

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

**Pull request merged**